stl/inc/memory

```diff
@@ -1578,18 +1851,309 @@ private:
     }
 };
 
+#ifdef __cpp_concepts
+template <class _Alloc>
+class _Uninitialized_rev_destroying_backout_al { // struct to undo partially constructed ranges in
+                                                 // _Uninitialized_xxx_al algorithms
+    using pointer = _Alloc_ptr_t<_Alloc>;
+
+public:
+    _Uninitialized_rev_destroying_backout_al(pointer _Dest, _Alloc& _Al_) : _First(_Dest), _Last(_Dest), _Al(_Al_) {}
+
+    _Uninitialized_rev_destroying_backout_al(const _Uninitialized_rev_destroying_backout_al&) = delete;
+    _Uninitialized_rev_destroying_backout_al& operator=(const _Uninitialized_rev_destroying_backout_al&) = delete;
+
+    ~_Uninitialized_rev_destroying_backout_al() {
+        while (_Last != _First) {
+            --_Last;
+            allocator_traits<_Alloc>::destroy(_Al, _Last);
+        }
+    }
+
+    template <class... _Types>
+    void _Emplace_back(_Types&&... _Vals) { // construct a new element at *_Last and increment
+        allocator_traits<_Alloc>::construct(_Al, _Unfancy(_Last), _STD forward<_Types>(_Vals)...);
+        ++_Last;
+    }
+
+    pointer _Release() { // suppress any exception handling backout and return _Last
+        _First = _Last;
+        return _Last;
+    }
+
+private:
+    pointer _First;
+    pointer _Last;
+    _Alloc& _Al;
+};
+
+template <class _Ty, class _Alloc>
+void _Reverse_destroy_multidimensional_n_al(_Ty* const _Arr, size_t _Size, _Alloc& _Al) {
+    while (_Size > 0) {
+        --_Size;
+        // TRANSITION, LLVM-42404
+        if constexpr ((rank_v<_Ty>) > 0) {
+            _Reverse_destroy_multidimensional_n_al(_Arr[_Size], extent_v<_Ty>, _Al);
+        } else {
+            allocator_traits<_Alloc>::destroy(_Al, _Arr + _Size);
+        }
+    }
+}
+
+template <class _Ty, size_t _Size, class _Alloc>
+void _Uninitialized_copy_multidimensional_al(const _Ty (&_In)[_Size], _Ty (&_Out)[_Size], _Alloc& _Al) {
+    // TRANSITION, LLVM-42404
+    if constexpr ((rank_v<_Ty>) > 0) {
+        for (size_t _Idx = 0; _Idx < _Size; ++_Idx) {
+            _TRY_BEGIN
+            _Uninitialized_copy_multidimensional_al(_In[_Idx], _Out[_Idx], _Al);
+            _CATCH_ALL
+            _Reverse_destroy_multidimensional_n_al(_Out, _Idx, _Al);
+            _RERAISE;
+            _CATCH_END
+        }
+    } else if constexpr (conjunction_v<bool_constant<_Ptr_copy_cat<_Ty*, _Ty*>::_Really_trivial>,
+                             _Uses_default_construct<_Alloc, _Ty*, _Ty>>) {
+        (void) _Al;

```

Comment:
Maybe one of maintainers could say which one is preferred.

Reply:
Our current convention is to use `(void)` casts to work around the compiler bug where variables are reported as unused when they're actually used in a non-taken branch of `if constexpr`. For variables that are unused depending on debug mode or other macros, old code uses `(void)` casts but new code sometimes uses `[[maybe_unused]]` - our conventions aren't settled there. (I think I like the attribute for "permanent" silencing in the presence of macros, since it's less verbose.)

Reply:
Ok, so I'll leave `(void)` cast.